### DISCUSSION TITLE
Problems adding timeout + timeout message to html IAT

### DISCUSSION THREAD

**Discussion by catpattie** [jspsych/jsPsych#1049] [Q&A]
Hello!

Just wondering if someone can help me. I have attempted to edit the iat-html plugin so that if the participant doesn't respond to a trial within 5 seconds it displays a timeout message and moves onto the next trial. I've successfully done this using the code below:

   // end trial if time limit is set 
    if (trial.trial_duration !== null && trial.response_ends_trial === true) {
      jsPsych.pluginAPI.setTimeout(function() {
        display_element.innerHTML += trial.timeout_message;
        jsPsych.pluginAPI.setTimeout(function() {
          end_trial();
        }, 2000);
        }, trial.trial_duration);
    }
  };

But for some reason, this is interacting weirdly with the display_feedback and html_when_wrong parameters. If I make an error on a trial and the feedback appears, then I let the trial timeout before selecting the correct answer, some subsequent correct answers are triggering the html_when_wrong, and I'm concerned that they'll be coded as incorrect responses. The only other part of the plugin I changed was adding 

timeout_message: {
        type: jsPsych.plugins.parameterType.STRING,
        pretty_name: 'Timeout message',
        default: "<p>Please respond faster.</p>",
        description: 'The message displayed on a timeout non-response.'
      },

to the parameters.

Thanks in advance if anyone is able to help!

Cat Pattie

**Comment by becky-gilbert** [comment]
Hi @catpattie! You were close. The reason this didn't work is because you set it up so that the timeout message and delay occurs on every trial, rather than only when there was no response before the `trial_duration`. So inside the first `setTimeout`, you just need to check whether a response was made - if so then it should call `end_trial`, and if not then you can add the `timeout_message` to the screen and set up another `setTimeout` to end the trial after that.
```javascript
if (trial.trial_duration !== null && trial.response_ends_trial != true) {
  jsPsych.pluginAPI.setTimeout(function() {
    if (response.rt == null) {
      display_element.innerHTML += trial.timeout_message;
      jsPsych.pluginAPI.setTimeout(function() {
        end_trial();
      }, 2000);
    } else {
      end_trial();
    }
  }, trial.trial_duration);
}
```
Another issue I saw was that the timeout message was causing the page content to jump. This occurs because there is content _below_ the timeout message, and so adding this message on to the page's HTML pushes the other content down. One solution to this is to re-use the div that contains the `html_when_wrong` content (i.e. the big red X). This container already exists on the page and its visibility is set to 'hidden'. So what you can do is put your timeout message inside of this div and change the visibility to 'visible'. The only reason you wouldn't want to do it this way is if you want to be able to show both the `html_when_wrong` and `timeout_message` at the same time.
```javascript
if (trial.trial_duration !== null && trial.response_ends_trial != true) {
  jsPsych.pluginAPI.setTimeout(function() {
    if (response.rt == null) {
      display_element.querySelector("#wrongImgContainer").innerHTML = trial.timeout_message;
      display_element.querySelector("#wrongImgContainer").style.visibility = "visible";
      jsPsych.pluginAPI.setTimeout(function() {
        end_trial();
      }, 2000);
    } else {
      end_trial();
    }
  }, trial.trial_duration);
}
```

**Reply by catpattie** [reply]
Thanks so much for your help Becky! I'll have another go. Promising to know I was close the first time too :)
